src/sentry/search/snuba/backend.py

```diff
@@ -231,76 +220,119 @@ def _query(self, project, retention_window_start, group_queryset, tags, environm
-            tags=tags,
-            start=start,
-            end=end,
-            sort=sort,
-            extra_aggregations=extra_aggregations,
-            score_fn=score_fn,
-            candidate_hashes=candidate_hashes,
-            **parameters
-        )
-        metrics.timing('snuba.search.num_snuba_results', len(snuba_groups))
+            if not snuba_groups:
+                break
 
-        if candidate_hashes:
-            # pre-filtered candidates were passed down to Snuba,
-            # so we're finished with filtering
-            result_groups = snuba_groups.items()
-        else:
-            # pre-filtered candidates were *not* passed down to Snuba,
-            # so we need to do post-filtering to verify Sentry DB predicates
-            result_groups = []
-            i = 0
-            for i, chunk in enumerate(chunked(snuba_groups.items(), MAX_POST_SNUBA_CHUNK), 1):
+            if candidate_hashes:
+                # pre-filtered candidates were passed down to Snuba,
+                # so we're finished with filtering and these are the
+                # only results
+                result_groups = snuba_groups
+            else:
+                # pre-filtered candidates were *not* passed down to Snuba,
+                # so we need to do post-filtering to verify Sentry DB predicates
                 filtered_group_ids = group_queryset.filter(
-                    id__in=[gid for gid, _ in chunk]
+                    id__in=[gid for gid, _ in snuba_groups]
                 ).values_list('id', flat=True)
 
-                result_groups.extend(
-                    (group_id, snuba_groups[group_id])
-                    for group_id in filtered_group_ids
-                )
-
-            metrics.timing('snuba.search.num_post_filters', i)
-
-        paginator_results = SequencePaginator(
-            [(score, id) for (id, score) in result_groups],
-            reverse=True,
-            **paginator_options
-        ).get_result(limit, cursor, count_hits=count_hits)
+                existing_group_ids = set(r[0] for r in result_groups)
+                group_to_score = dict(snuba_groups)
+                for group_id in filtered_group_ids:
+                    if group_id in existing_group_ids:
+                        # because we're doing multiple Snuba queries, which
+                        # happen outside a transaction, there is a small possibility
+                        # of groups moving around in the sort scoring underneath us,
+                        # so we at least want to protect against duplicates
+                        continue
+
+                    result_groups.append((group_id, group_to_score[group_id]))
+
+            paginator_results = SequencePaginator(
+                [(score, id) for (id, score) in result_groups],
+                reverse=True,
+                **paginator_options
+            ).get_result(limit, cursor, count_hits=count_hits)
```

Comment:
I have to figure out/remember what `count_hits` is for because it basically won't work anymore...

Reply:
It's the number shown next to the saved search dropdown/heading:

![screenshot 2018-09-17 17 08 19](https://user-images.githubusercontent.com/65315/45656844-868f4180-ba9c-11e8-802e-9ceb0ea7ccc6.png)

Reply:
Yeah I literally dunno how we could handle that and not do the current "pass the entire contents of one DB to the other DB."

I'll try to brain it more tomorrow morning but open to idears.